rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+                r = self.listener_6.notified() => r?,
+            }
+            Ok(())
+        }
+    }
+
+    struct Listener {
+        key: winreg::RegKey,
+        notify: std::pin::Pin<Box<Notify>>,
+        wait_handle: Option<HANDLE>,
+        event: Option<HANDLE>,
+    }
+
+    impl Listener {
+        pub(crate) fn new(key: winreg::RegKey) -> Result<Self> {
+            let notify = Box::pin(Notify::default());
+            let notify_ptr: *const Notify = notify.deref();
+            let event = unsafe { CreateEventA(None, false, false, None) }?;
+            let mut wait_handle = HANDLE(0isize);
+
+            // The docs say that `RegisterWaitForSingleObject` uses "a worker thread" from
+            // "the thread pool".
+            // Raymond Chen, who is an authority on Windows internals, says RegisterWaitForSingleObject
+            // does multiple waits on the same worker thread internally:
+            // <https://devblogs.microsoft.com/oldnewthing/20081117-00/?p=20183>
+            // There is a different function in the CLR called `RegisterWaitForSingleObject`,
+            // so he might be talking about that.
+            // If he's not, then we're invisibly tying up two worker threads. Can't help it.
+
+            // SAFETY: It's complicated.
+            // The callback here can cause a lot of problems. We pin the `Notify` object.
+            // We don't use an `Arc` because `Notify` already has only `&self` methods, and
+            // the callback has no way to free an `Arc`.
+            // When we call `UnregisterWaitEx` later, we wait for all callbacks to finish running
+            // before we drop the `Notify`, to avoid a dangling pointer.
+            unsafe {
+                RegisterWaitForSingleObject(
+                    &mut wait_handle,
+                    event,
+                    Some(callback),
+                    Some(notify_ptr as *const _),
+                    INFINITE,
+                    WT_EXECUTEINWAITTHREAD,
+                )
+            }?;
+
+            let mut that = Self {
+                key,
+                notify,
+                wait_handle: Some(wait_handle),
+                event: Some(event),
+            };
+            that.register_callback()?;
+
+            Ok(that)
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
+            if let Some(wait_handle) = self.wait_handle.take() {
+                unsafe { UnregisterWaitEx(wait_handle, INVALID_HANDLE_VALUE) }?;
+            }
+            if let Some(event) = self.event.take() {
+                unsafe { CloseHandle(event) }?;
+            }
+
+            Ok(())
+        }
+
+        // I'm not sure if this needs to be mut, I'm being cautious here.
+        // There is no reason in practice you'd ever want to call it from two
+        // threads at once.
+        //
+        // Cancel-safety: Probably. `tokio::sync::Notify` sounds cancel-safe.
```

Comment:
[it's not](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) we should use `mpsc::channel` with buffer size 1 if we want cancellation safety.

Reply:
Darn. I thought "lose your place in the queue" would be irrelevant if there's only one waiter.
Done